daq/runner.py

```diff
@@ -587,46 +592,67 @@ def main_loop(self):
 
         self._terminate()
 
-    def _target_set_check_state(self, device, remote_trigger):
-        assert self._devices.contains(device), 'Target device %s is not expected' % device.mac
+    def _target_set_trigger(self, device, remote_trigger=False):
+        assert self._devices.contains(device), 'Target device %s is not expected' % device
+
         if not self._system_active:
-            LOGGER.warning('Target device %s ignored, system is not active', device.mac)
-            return False
+            LOGGER.warning('Target device %s ignored, system is not active', device)
+            return
 
         if device.host:
-            LOGGER.debug('Target device %s already triggered', device.mac)
-            return False
+            LOGGER.debug('Target device %s already triggered', device)
+            return
 
         if not self.run_tests:
-            LOGGER.debug('Target device %s trigger suppressed', device.mac)
-            return False
+            LOGGER.debug('Target device %s trigger suppressed', device)
+            return
 
         if device.wait_remote and not remote_trigger:
             LOGGER.debug('Ignoring local trigger for remote target %s', device)
-            return False
+            return
 
-        return True
+        if device.queued:
+            LOGGER.debug('Target device %s already queued', device)
+            return
 
-    def _target_set_trigger(self, device, remote_trigger=False):
-        if not self._target_set_check_state(device, remote_trigger):
-            return False
+        device.wait_remote = False
+        LOGGER.info('TAPTAP queue device %s', device)
+        device.queued = True
+        self._target_set_queue.put(device)
+        if self._target_triggered_maxed(additional=self._target_set_queue.qsize()):
+            LOGGER.info('Target device %s triggering queued (%s)',
+                        device, self._target_set_queue.qsize())
+
+    def _target_triggered_maxed(self, additional=0):
+        num_triggered = len(self._devices.get_triggered_devices())
+        return self._concurrent_runs > 0 and (num_triggered + additional) >= self._concurrent_runs
+
+    def _target_set_consider(self):
+        if self._target_set_queue.empty():
+            return
+        if self._target_triggered_maxed():
+            return
+        device = self._target_set_queue.get()
+        if device.queued:
```

Comment:
Is there a case where device.queued is false when it's in the queue? otherwise, having device in _target_set_queue and device.queued are duplicated information.

Reply:
If the trigger is canceled for whatever reason while it's still in the queue then this can happen. Partly this is because Queue is kinda lame and you can't remove something that's in it already.  Another alternative is I could use a regular list and just append and pop(0), which are functionally the same, because then if the device is still in the queue when it's cancelled it could just be removed.... ?

Reply:
cleaning this up to use a standard array -- much simpler overall